### PR TITLE
feat: add replay syncer and get replay status

### PR DESCRIPTION
**Pull request by arinda-arif**

### CODE REVIEW

File: cmd/replay.go [hunk 1]
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
...
+	reCmd := &cli.Command{
+		Use:     "status",
+		Short:   "get status of a replay using its ID",
+		Example: "optimus replay status replay-id",
+		Long: `
+This operation takes one argument, replay ID[required]
+that generated when starting a replay.
+		`,
+		Args: func(cmd *cli.Command, args []string) error {
+			if len(args) < 2 {
```
Comment (kushsharma) [inline cmd/replay.go:289]: In the description, you said we need 1 argument, here we are looking for 2.

File: cmd/replay.go [hunk 2]
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
...
+				return errors.New("job name and replay ID are required")
```
Comment (kushsharma) [inline cmd/replay.go:290]: Btw, isn't replay id unique, why do we need a job name?

File: cmd/server/server.go [hunk 3]
```diff
@@ -314,7 +314,8 @@ func Initialize(conf config.Provider) error {
 		db:   dbConn,
 		hash: appHash,
 	}
-	registeredProjects, err := projectRepoFac.New().GetAll()
+	projectRepo := projectRepoFac.New()
```
Comment (kushsharma) [inline cmd/server/server.go:317]: why break this into 2 lines?

File: job/replay_manager.go [hunk 4]
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
 
 type ReplayManager interface {
 	Init()
-	Replay(context.Context, *models.ReplayWorkerRequest) (string, error)
+	Replay(context.Context, *models.ReplayRequest) (string, error)
+	GetReplay(uuid uuid.UUID) (*models.ReplaySpec, error)
+	GetRunsStatus(ctx context.Context, reqInput *models.ReplayRequest, jobName string) ([]models.JobStatus, error)
```
Comment (kushsharma) [inline job/replay_manager.go:45]: I think `GetRunStatus` sounds better

File: job/replay_manager.go [hunk 5]
```diff
@@ -236,6 +133,18 @@ func (m *Manager) spawnServiceWorker() {
...
+// AirflowSyncer to sync for latest replay status
+func (m *Manager) AirflowSyncer() {
+	logger.I("start synchronizing replays...")
+	ctx, cancelCtx := context.WithTimeout(context.Background(), m.config.WorkerTimeout)
+	if err := m.replaySyncer.Sync(ctx, m.config.RunTimeout); err != nil {
+		logger.E(errors.Wrap(err, "syncer failed to process"))
+		cancelCtx()
+	}
+	logger.I("replays synced")
+	cancelCtx()
```
Comment (kushsharma) [inline job/replay_manager.go:145]: We can `defer` this call before line 140 right to avoid calling it at 2 places.

**Review by kushsharma — CHANGES_REQUESTED** [review]

File: api/handler/v1/runtime.go [hunk 6]
```diff
@@ -803,23 +806,60 @@ func (sv *RuntimeServiceServer) Replay(ctx context.Context, req *pb.ReplayReques
...
-		return nil, status.Errorf(codes.NotFound, "%s: project %s not found", err.Error(), req.GetProjectName())
+		return nil, err
 	}
 
-	namespaceRepo := sv.namespaceRepoFactory.New(projSpec)
-	namespaceSpec, err := namespaceRepo.GetByName(req.GetNamespace())
+	replayState, err := sv.jobSvc.GetStatus(ctx, replayRequest)
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "%s: namespace %s not found", err.Error(), req.GetNamespace())
+		return nil, err
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply api/handler/v1/runtime.go:817]: yes, but for this i have set the status.Error inside the called function (getProjectSpec and getJobSpec). let me know if this is make sense or I should only return the error message and wrap the status.Err in the line you mentioned

File: cmd/replay.go [hunk 7]
```diff
@@ -267,3 +270,90 @@ func runReplayRequest(l logger, projectName, namespace, jobName, startDate, endD
...
+		namespace     string
+	)
+
+	reCmd := &cli.Command{
+		Use:     "status",
+		Short:   "get status of a replay using its ID",
+		Example: "optimus replay status replay-id",
+		Long: `
+This operation takes one argument, replay ID[required]
+that generated when starting a replay.
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply cmd/replay.go:286]: got it

File: cmd/replay.go (hunk 2 above)
Comment (arinda-arif) [reply cmd/replay.go:290]: yes, it is unique. have added it for the api call: https://github.com/odpf/proton/blob/4bcadc9389fa7e895509fa85cd7a2c25af5f3dc6/odpf/optimus/runtime_service.proto#L192
let me know your thoughts and if you think removing it is more proper

File: api/handler/v1/runtime.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kushsharma) [reply api/handler/v1/runtime.go:817]: I don't see any of these functions, I see `replayState, err := sv.jobSvc.GetStatus(ctx, replayRequest)`

File: cmd/replay.go (hunk 2 above)
Comment (kushsharma) [reply cmd/replay.go:290]: For rest call to make it follow the convention I guess we need it there but we are using GRPC here, do we really need it?

File: api/handler/v1/runtime.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arinda-arif) [reply api/handler/v1/runtime.go]: updated

File: cmd/replay.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (arinda-arif) [reply cmd/replay.go]: updated

File: cmd/replay.go (hunk 1 above)
Comment (arinda-arif) [reply cmd/replay.go]: updated to only 1

File: cmd/replay.go (hunk 2 above)
Comment (arinda-arif) [reply cmd/replay.go]: agreed. updated to only require replay ID here

File: cmd/server/server.go (hunk 3 above)
Comment (arinda-arif) [reply cmd/server/server.go]: merged into 1

File: job/replay_manager.go (hunk 4 above)
Comment (arinda-arif) [reply job/replay_manager.go]: updated

File: job/replay_manager.go (hunk 5 above)
Comment (arinda-arif) [reply job/replay_manager.go]: changed to use defer

File: job/replay_manager.go [hunk 8]
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
 
 type ReplayManager interface {
 	Init()
-	Replay(context.Context, *models.ReplayWorkerRequest) (string, error)
+	Replay(context.Context, *models.ReplayRequest) (string, error)
```
Comment (kushsharma) [inline job/replay_manager.go:43]: Any reason we are using pointer in the request?

File: job/replay_manager.go [hunk 9]
```diff
@@ -37,7 +40,9 @@ type ReplayManagerConfig struct {
...
+	GetReplay(uuid uuid.UUID) (*models.ReplaySpec, error)
```
Comment (kushsharma) [inline job/replay_manager.go:44]: Any reason we are using pointer here as well?

File: job/replay_syncer.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	replaySpecFactory         ReplaySpecRepoFactory
+	scheduler                 models.SchedulerUnit
+	dependencyResolver        DependencyResolver
+	projectJobSpecRepoFactory ProjectJobSpecRepoFactory
+	assetCompiler             AssetCompiler
+	Now                       func() time.Time
+	projects                  []models.ProjectSpec
+}
+
+func NewReplaySyncer(replaySpecFactory ReplaySpecRepoFactory, scheduler models.SchedulerUnit, dependencyResolver DependencyResolver, projectJobSpecRepoFactory ProjectJobSpecRepoFactory, assetCompiler AssetCompiler, projects []models.ProjectSpec) *Syncer {
```
Comment (kushsharma) [inline job/replay_syncer.go:36]: We can format these lines to have a soft limit to 80 and a hard limit to 120 chars

File: job/replay_syncer.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+	return &Syncer{replaySpecFactory: replaySpecFactory, scheduler: scheduler, dependencyResolver: dependencyResolver, projectJobSpecRepoFactory: projectJobSpecRepoFactory, assetCompiler: assetCompiler, Now: time.Now, projects: projects}
```
Comment (kushsharma) [inline job/replay_syncer.go:37]: time should be injected in the constructor and with UTC

File: job/replay_syncer.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+func (s Syncer) Sync(context context.Context, runTimeout time.Duration) error {
+	replaySpecRepo := s.replaySpecFactory.New(models.JobSpec{})
```
Comment (kushsharma) [inline job/replay_syncer.go:41]: Why do we need to pass this empty jobspec?

File: models/job.go [hunk 13]
```diff
@@ -58,6 +58,7 @@ type JobSpec struct {
 	Dependencies map[string]JobSpecDependency // job name to dependency
 	Assets       JobAssets
 	Hooks        []JobSpecHook
+	Project      ProjectSpec
```
Comment (kushsharma) [inline models/job.go:61]: Why is this added?

File: job/replay_syncer.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+type Syncer struct {
+	replaySpecFactory         ReplaySpecRepoFactory
+	scheduler                 models.SchedulerUnit
+	dependencyResolver        DependencyResolver
+	projectJobSpecRepoFactory ProjectJobSpecRepoFactory
+	assetCompiler             AssetCompiler
+	Now                       func() time.Time
+	projects                  []models.ProjectSpec
```
Comment (kushsharma) [inline job/replay_syncer.go:33]: why inject them? Why not just fetch them from the repository when needed?

File: job/replay_syncer.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+		return nil, errors.Wrapf(err, "failed to retrieve jobs")
+	}
+
+	for i, jSpec := range jobSpecs {
+		if jobSpecs[i].Assets, err = s.assetCompiler(jSpec, s.Now()); err != nil {
+			return nil, errors.Wrap(err, "asset compilation")
+		}
+	}
+
+	var resolvedSpecs []models.JobSpec
```
Comment (kushsharma) [inline job/replay_syncer.go:129]: you are not really using this variable anywhere

File: job/replay_syncer.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+func (s Syncer) syncRunningReplay(context context.Context, replaySpec models.ReplaySpec, replaySpecRepo store.ReplaySpecRepository) error {
+	var projSpec models.ProjectSpec
+	for _, project := range s.projects {
+		if project.Name == replaySpec.Job.Project.Name {
+			projSpec = project
+			break
+		}
+	}
+	jobSpecMap, err := s.prepareJobSpecMap(projSpec)
```
Comment (kushsharma) [inline job/replay_syncer.go:88]: I think we store the execution tree in db, why do we need to create the tree again from scratch?

File: Makefile [hunk 17]
```diff
@@ -27,7 +27,7 @@ pack-files:
 generate-proto: ## regenerate protos
 	@echo " > cloning protobuf from odpf/proton"
 	@rm -rf proton/
-	@git -c advice.detachedHead=false clone https://github.com/odpf/proton --depth 1 --quiet --branch main
+	@git -c advice.detachedHead=false clone https://github.com/odpf/proton --depth 1 --quiet --branch optimus-replay-status
```
Comment (kushsharma) [inline Makefile:30]: I have merged the PR, you can change it to `main`

File: job/replay_syncer.go [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	//for each replay, will get its list of jobs and runs and check if completed
+	treeNode, err := prepareReplayExecutionTree(replayRequest)
+	if err != nil {
+		return err
+	}
+	allNodes := treeNode.GetAllNodes()
+
+	stateSummary, err := s.checkInstanceState(context, projSpec, allNodes, replayRequest.Start, replayRequest.End)
```
Comment (kushsharma) [inline job/replay_syncer.go:117]: why not use `treeNode.GetAllNodes()` directly here?

File: job/replay_syncer.go [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+		}
+	} else if stateSummary[models.InstanceStateRunning] == 0 && stateSummary[models.InstanceStateFailed] == 0 && stateSummary[models.InstanceStateSuccess] > 0 {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replayID, models.ReplayStatusSuccess, models.ReplayMessage{
+			Type:    models.ReplayStatusSuccess,
+			Message: ReplayMessageSuccess,
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking replay as success error: %s", updateStatusErr))
+		}
+	}
+	return nil
```
Comment (kushsharma) [inline job/replay_syncer.go:189]: I think you need to return error here `updateStatusErr `

File: job/replay_manager.go (hunk 8 above)
Comment (arinda-arif) [reply job/replay_manager.go]: was read somewhere that instead of passing the whole struct, better practice is to reference it. let me know your thoughts & consideration?

File: job/replay_syncer.go (hunk 11 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: do this mean 'Now' value should also passed to the constructor, not being initialized in the constructor?

File: job/replay_syncer.go (hunk 12 above)
Comment (arinda-arif) [reply job/replay_syncer.go:50]: initially we are constructing ReplayRepository with a JobSpec, but just realized now that we don't really need it because of the latest changes. will update

File: models/job.go (hunk 13 above)
Comment (arinda-arif) [reply models/job.go:61]: have added this because when getting Replay from the repository, need to have Job and Project. second thought, i will remove this and use project from user input instead

File: job/replay_syncer.go (hunk 16 above)
Comment (arinda-arif) [reply job/replay_syncer.go:97]: no we are not storing the replay tree in db, but i missed the discussion on why we conclude on this. thoughts?

File: job/replay_syncer.go (hunk 14 above)
Comment (arinda-arif) [reply job/replay_syncer.go:33]: fixing this

File: models/job.go (hunk 13 above)
Comment (arinda-arif) [reply models/job.go:61]: another usage: on the replay syncer, we need to know the project of each active replays that need to be synced, so in the replay repository, we fetch Replay that got preloaded with Job and Project (that is why Project is added). Thoughts?

File: job/replay_syncer.go (hunk 10 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: i see you have refactored this, thanks!

File: job/replay_syncer.go (hunk 11 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: done by Kush

File: job/replay_syncer.go (hunk 12 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: updated. not passing jobspec to constructor

File: job/replay_syncer.go (hunk 14 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: changed []models.ProjectSpec to ProjectRepoFactory

File: job/replay_syncer.go (hunk 15 above)
Comment (arinda-arif) [reply job/replay_syncer.go:134]: fixed

File: Makefile (hunk 17 above)
Comment (arinda-arif) [reply Makefile]: updated

File: job/replay_syncer.go (hunk 18 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_syncer.go (hunk 19 above)
Comment (arinda-arif) [reply job/replay_syncer.go:187]: added the return

File: job/replay_manager.go (hunk 8 above)
Comment (kushsharma) [reply job/replay_manager.go]: If the struct is large then only it makes sense, in these cases, to pass it via reference, else it is actually bad for cache coherence. CPU works with L1, L2, L3... caches, each of them has different sizes, it's actually faster to copy a small struct and keep it in the cache then to dereference a pointer.

Comment (arinda-arif) [reply job/replay_manager.go]: Thank you for the explanation. I have updated to not use pointer in replay requests.

File: job/replay_manager.go (hunk 9 above)
Comment (arinda-arif) [reply job/replay_manager.go]: Have updated this too to not use pointer

File: job/replay_syncer.go (hunk 16 above)
Comment (arinda-arif) [reply job/replay_syncer.go]: execution tree is now stored in db

**Comment by arinda-arif** [comment]
Hi @kushsharma ,

I have made several changes, please help to check:
- Storing execution tree in DB (in replay manager), and fetch it when being used in replay worker and syncer.
- Queue to be taken by replay worker is now only represented by uuid only, since in replay worker we are fetching from db anyway.
- Not use pointers in ReplayRequest and ReplaySpec objects.
- Getting replay from repository is now preloaded with project and secret.

File: job/service.go [hunk 20]
```diff
@@ -54,7 +55,19 @@ type JobRepoFactory interface {
...
+// ProjectRepoFactory is used to manage projects from store
+type ProjectRepoFactory interface {
+	New() store.ProjectRepository
+}
+
+type ReplayManager interface {
+	Init()
+	Replay(context.Context, models.ReplayRequest) (string, error)
+	GetReplay(uuid uuid.UUID) (models.ReplaySpec, error)
+	GetRunStatus(ctx context.Context, replaySpec models.ReplaySpec, jobSpec models.JobSpec) ([]models.JobStatus, error)
```
Comment (kushsharma) [inline job/service.go:70]: I see replay spec already has job spec then why are we explicitly passing it again as an argument?

File: store/postgres/replay_repository.go [hunk 21]
```diff
@@ -19,57 +22,112 @@ type Replay struct {
...
+		EndDate:       spec.EndDate.UTC(),
+		Status:        spec.Status,
+		Message:       message,
+		ExecutionTree: executionTree,
 	}, nil
 }
 
+func toTreeNode(executionTree *ExecutionTree) *tree.TreeNode {
+	var jobSpec models.JobSpec
+	mapstructure.Decode(executionTree.Data, &jobSpec)
```
Comment (kushsharma) [inline store/postgres/replay_repository.go:86]: why are we ignoring error returned by this `Decode`?

File: store/postgres/replay_repository.go [hunk 22]
```diff
@@ -83,13 +141,23 @@ func (repo *replayRepository) Insert(replay *models.ReplaySpec) error {
 
 func (repo *replayRepository) GetByID(id uuid.UUID) (models.ReplaySpec, error) {
 	var r Replay
-	if err := repo.DB.Where("id = ?", id).Find(&r).Error; err != nil {
+	if err := repo.DB.Where("id = ?", id).Preload("Job").Preload("Job.Project").
+		Preload("Job.Project.Secrets").Find(&r).Error; err != nil {
```
Comment (kushsharma) [inline store/postgres/replay_repository.go:145]: Can we skip loading Projects & secrets? Sorry, I can't find where we are using them?

**Review by kushsharma — CHANGES_REQUESTED** [review]

File: store/postgres/replay_repository.go (hunk 22 above)
Comment (arinda-arif) [reply store/postgres/replay_repository.go:145]: we are using them in scheduler, since we need scheduler auth token and scheduler host to get dag run status. should i just load them just before calling the scheduler then? in replay manager?

File: models/job.go (hunk 13 above)
Comment (arinda-arif) [reply models/job.go:61]: yes, i can remove this and call project where it is needed (when calling scheduler)

File: store/postgres/replay_repository.go (hunk 22 above)
Comment (arinda-arif) [reply store/postgres/replay_repository.go]: updated to not preload projects and secrets, since in replay syncer we fetched all the projects and when getting replay, we are getting it from the user

File: store/postgres/replay_repository.go (hunk 21 above)
Comment (arinda-arif) [reply store/postgres/replay_repository.go]: refactored to not use mapstructure

File: job/service.go (hunk 20 above)
Comment (arinda-arif) [reply job/service.go]: updated to not use replaySpec and replaced with the necessary fields to avoid confusion

File: job/replay_syncer.go [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+	stateSummary[models.InstanceStateSuccess] = 0
+
+	for _, node := range replaySpec.ExecutionTree.GetAllNodes() {
+		batchEndDate := replaySpec.EndDate.AddDate(0, 0, 1).Add(time.Second * -1)
+		jobStatusAllRuns, err := s.scheduler.GetDagRunStatus(ctx, node.Data.(models.JobSpec), replaySpec.StartDate, batchEndDate, schedulerBatchSize)
+		if err != nil {
+			return nil, err
+		}
+		for _, jobStatus := range jobStatusAllRuns {
+			stateSummary[jobStatus.State.String()] = stateSummary[jobStatus.State.String()] + 1
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_syncer.go [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+	runningTime := time.Now().Sub(replaySpec.CreatedAt)
+	if runningTime > runTimeout {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replaySpec.ID, models.ReplayStatusFailed, models.ReplayMessage{
+			Type:    ReplayRunTimeout,
+			Message: fmt.Sprintf("replay has been running since %s", replaySpec.CreatedAt.UTC().Format(TimestampLogFormat)),
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking long running replay jobs as failed: %s", updateStatusErr))
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go:86]: updated

File: job/replay_syncer.go [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+		// sync timed out replays for accepted and in progress replays
+		if err := syncTimedOutReplay(replaySpecRepo, replaySpec, runTimeout); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func syncTimedOutReplay(replaySpecRepo store.ReplaySpecRepository, replaySpec models.ReplaySpec, runTimeout time.Duration) error {
+	runningTime := time.Now().Sub(replaySpec.CreatedAt)
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_manager.go [hunk 26]
```diff
@@ -236,6 +131,17 @@ func (m *Manager) spawnServiceWorker() {
 	}
 }
 
+// AirflowSyncer to sync for latest replay status
+func (m *Manager) AirflowSyncer() {
+	logger.I("start synchronizing replays...")
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_manager.go]: updated

File: cmd/server/server.go [hunk 27]
```diff
@@ -79,10 +79,11 @@ func (fac *projectJobSpecRepoFactory) New(project models.ProjectSpec) store.Proj
 type replaySpecRepoRepository struct {
 	db             *gorm.DB
 	jobSpecRepoFac jobSpecRepoFactory
+	hash           models.ApplicationKey
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply cmd/server/server.go]: updated, no longer need hash since we are not preload secrets in replay repo

File: api/handler/v1/adapter_test.go [hunk 28]
```diff
@@ -29,6 +32,25 @@ func TestAdapter(t *testing.T) {
...
+			ScheduledAt: timeRun,
+		}
+		treeNode.Runs = set.NewTreeSetWith(job.TimeOfJobStatusComparator)
+		treeNode.Runs.Add(jobStatus)
+		adap := v1.Adapter{}
+		replayExecutionTreeNode, err := adap.ToReplayStatusTreeNode(treeNode)
+		assert.Nil(t, err)
+		assert.Equal(t, replayExecutionTreeNode.JobName, "job-name")
+		assert.Equal(t, 1, len(replayExecutionTreeNode.Dependents))
+		assert.Equal(t, replayExecutionTreeNode.Dependents[0].JobName, "nested-job-name")
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply api/handler/v1/adapter_test.go]: updated

File: api/handler/v1/runtime.go [hunk 29]
```diff
@@ -803,47 +806,85 @@ func (sv *RuntimeServiceServer) Replay(ctx context.Context, req *pb.ReplayReques
...
 }
 
-func (sv *RuntimeServiceServer) parseReplayRequest(req *pb.ReplayRequest) (*models.ReplayWorkerRequest, error) {
-	projectRepo := sv.projectRepoFactory.New()
-	projSpec, err := projectRepo.GetByName(req.GetProjectName())
+func (sv *RuntimeServiceServer) GetReplayStatus(ctx context.Context, req *pb.GetReplayStatusRequest) (*pb.GetReplayStatusResponse, error) {
+	uuid, err := uuid.Parse(req.Id)
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "%s: project %s not found", err.Error(), req.GetProjectName())
+		return nil, status.Errorf(codes.Internal, "error while getting replay: %v", err)
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply api/handler/v1/runtime.go]: updated

File: job/replay_validator.go [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+var (
+	// ReplayStatusToValidate signifies list of status to be used when checking active replays
+	ReplayStatusToValidate = []string{models.ReplayStatusInProgress, models.ReplayStatusAccepted, models.ReplayStatusReplayed}
+)
+
+type ReplayValidator interface {
+	Validate(context.Context, store.ReplaySpecRepository, models.ReplayRequest, *tree.TreeNode) error
+}
```
Comment (kushsharma) [inline job/replay_validator.go:20]: We should move this to `replay_manager.go`

File: job/replay_validator.go [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+			End:        activeSpec.EndDate,
+			Project:    reqInput.Project,
+			JobSpecMap: reqInput.JobSpecMap,
+		}
+		activeTree, err := prepareReplayExecutionTree(activeReplayWorkerRequest)
+		if err != nil {
+			return err
+		}
+		activeNodes := activeTree.GetAllNodes()
+		return checkAnyConflictedDags(activeNodes, reqReplayNodes)
```
Comment (kushsharma) [inline job/replay_validator.go:105]: This for loop will run only once instead of iterating over all `activeReplaySpecs`. Is this expected?

**Review by kushsharma — CHANGES_REQUESTED** [review]

File: job/replay.go [hunk 32]
```diff
@@ -16,37 +18,40 @@ const (
...
-	if err := srv.populateRequestWithJobSpecs(replayRequest); err != nil {
+func (srv *Service) ReplayDryRun(replayRequest models.ReplayRequest) (*tree.TreeNode, error) {
+	jobSpecMap, err := srv.prepareJobSpecMap(replayRequest)
+	if err != nil {
 		return nil, err
 	}
+	replayRequest.JobSpecMap = jobSpecMap
 
-	rootInstance, err := prepareTree(replayRequest)
+	rootInstance, err := prepareReplayExecutionTree(replayRequest)
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay.go]: updated

File: job/replay_syncer.go [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+
+var (
+	ReplayStatusToSynced = []string{models.ReplayStatusReplayed, models.ReplayStatusInProgress, models.ReplayStatusAccepted}
+	ReplayMessageSuccess = "all instances for this replay are successfully run"
+	ReplayMessageFailed  = "instance run failure found"
+)
+
+type ReplaySyncer interface {
+	Sync(context.Context, time.Duration) error
+}
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: moved

File: job/replay_syncer.go [hunk 34]
```diff
@@ -0,0 +1,136 @@
...
+
+func (s Syncer) syncTimedOutReplay(replaySpecRepo store.ReplaySpecRepository, replaySpec models.ReplaySpec, runTimeout time.Duration) error {
+	runningTime := s.Now().Sub(replaySpec.CreatedAt)
+	if runningTime > runTimeout {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replaySpec.ID, models.ReplayStatusFailed, models.ReplayMessage{
+			Type:    ReplayRunTimeout,
+			Message: fmt.Sprintf("replay has been running since %s", replaySpec.CreatedAt.UTC().Format(TimestampLogFormat)),
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking long running replay jobs as failed: %s", updateStatusErr))
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_syncer.go [hunk 35]
```diff
@@ -0,0 +1,136 @@
...
+	return stateSummary, nil
+}
+
+func updateCompletedReplays(stateSummary map[string]int, replaySpecRepo store.ReplaySpecRepository, replayID uuid.UUID) error {
+	if stateSummary[models.InstanceStateRunning] == 0 && stateSummary[models.InstanceStateFailed] > 0 {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replayID, models.ReplayStatusFailed, models.ReplayMessage{
+			Type:    models.ReplayStatusFailed,
+			Message: ReplayMessageFailed,
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking replay as failed error: %s", updateStatusErr))
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_syncer.go [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+			return updateStatusErr
+		}
+	} else if stateSummary[models.InstanceStateRunning] == 0 && stateSummary[models.InstanceStateFailed] == 0 && stateSummary[models.InstanceStateSuccess] > 0 {
+		if updateStatusErr := replaySpecRepo.UpdateStatus(replayID, models.ReplayStatusSuccess, models.ReplayMessage{
+			Type:    models.ReplayStatusSuccess,
+			Message: ReplayMessageSuccess,
+		}); updateStatusErr != nil {
+			logger.I(fmt.Sprintf("marking replay as success error: %s", updateStatusErr))
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply job/replay_syncer.go]: updated

File: job/replay_validator.go (hunk 30 above)
Comment (arinda-arif) [reply job/replay_validator.go]: moved

File: job/replay_validator.go (hunk 31 above)
Comment (arinda-arif) [reply job/replay_validator.go]: have updated this + added a test case to check this

**Comment by arinda-arif** [comment]
Hi @kushsharma, I have updated as the feedback. what I also added:
- start & finish syncer message log are now all in Debug mode (logger.D)
- added log info when a replay has been marked as success in syncer
- Process function in Replay Mock is now waiting for 200ms instead of 100ms, due to intermittent test error

**Comment by kushsharma** [comment]
I have removed that sleep completely and emulating it with channels.
